### PR TITLE
feat: detect when the resolved torch/CUDA ships no kernels for the local GPU

### PR DESCRIPTION
**Pull request by golithe**
- Cog resolves CUDA from the framework pin and never consults the GPU it runs on
- On a Blackwell card (sm_120), `torch==2.4.1` builds green, and every kernel launch fails with `CUDA error: no kernel image is available for execution on the device`
- Hard to notice: `torch.cuda.is_available()` returns `True`, and PyTorch's own diagnostic is a `UserWarning`
- There's no fallback either, since the wheel ships neither sm_12x cubins nor PTX to JIT from
- #389 is the same failure class on an A100 in 2022 (closed when the reporter worked around it); Blackwell makes it current again

This PR adds: 
- `GPUCompatibilityCheck` to `cog doctor`, modelled on `DockerCheck`
-  Reads compute capability from `nvidia-smi --query-gpu=compute_cap` (every distinct GPU, not just the weakest, since a mixed-GPU host must clear each device's floor) and compares the resolved torch/CUDA against the oldest release known to ship kernels for it.

`cog doctor` output when it fires:
  ```
   ⚙  Environment
   ✔ Docker
   ✔ Python version
   ⚠ GPU compatibility
   ⚙    torch==2.4.1 (CUDA 12.4) ships no kernels for sm_120, the compute capability of
        this machine's GPU. The image will build, but every CUDA operation in it will
        fail at runtime with "no kernel image is available for execution on the device".

   ⚙  Found 1 warning.
  ```
Notes for review:

- The floors are measured, not read off release notes. I installed each wheel and read `torch._C._cuda_getArchFlags()`, which works without a GPU
- Each floor is bracketed: the named version ships the kernels, the release below it doesn't. sm_120/sm_100 need torch>=2.7.0 + CUDA>=12.8; sm_90 needs torch>=2.0.1 + CUDA>=11.8
- Both bounds are load-bearing: a 2.7 build against an old CUDA (cu118) has no Blackwell kernels. The CUDA compared is the one from the wheel Cog actually resolves and installs (same resolution as `PythonRequirementsForArch`), not the raw requirement, so it's correct in both directions: `torch==2.7.0+cu118` with `cuda: "12.8"` resolves to the `cu128` wheel and stays silent, while an explicit `--extra-index-url=.../cu118` installs the cu118 wheel regardless of `cuda:` and warns against 11.8
- Deliberately no rows below sm_90. Every probed wheel already covers Turing/Ampere/Ada, so those floors can't be bracketed; unknown-old devices produce no finding rather than a guess
- `SeverityWarning`, matching `PythonVersionCheck`: the image is valid and runs fine on other hardware; it only fails when executed on this machine's GPU
- Silent when there's no GPU, no torch pin, or `gpu: false`; `COG_SKIP_GPU_CHECK=1` skips it when building for different hardware than the local card. Doctor-only, no build-path changes
- The torch pin is matched PEP 503-style, so `Torch==` and `torch[extra]==` resolve rather than silently skipping the check.

Testing:

- the comparison is a pure function (`evaluateGPUCompat`), table-tested without a GPU, including `+cu128` local-tag pins and the new-torch/old-CUDA case
- Verified by hand on an RTX 5070 Ti (sm_120): fires on `torch==2.4.1`, silent on `2.7.1` and on `gpu: false`

Out of scope (deliberately): 
- the floors are hand-entered from measurement (precedent: `MinimumTorchVersion` and friends in `pkg/dockerfile/base.go`)
- A probe script in `tools/` to regenerate them on demand could be a follow-up; full `compatgen` integration seems like a poor fit, since new capability majors ship every couple of years and the probe needs multi-GB wheel installs. Happy to add the script here or in a follow-up PR if useful

### CODE REVIEW

File: pkg/doctor/check_env_gpu_compat.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		return [2]int{}, false
+	}
+
+	sort.Slice(caps, func(i, j int) bool {
+		if caps[i][0] != caps[j][0] {
+			return caps[i][0] < caps[j][0]
+		}
+		return caps[i][1] < caps[j][1]
+	})
+	return caps[0], true
```
Comment (anish-sahoo) [inline pkg/doctor/check_env_gpu_compat.go]: Could we evaluate every detected capability, or select the capability with the strictest applicable floor? Sorting ascending and returning only `caps[0]` checks the numerically lowest GPU, which is not necessarily the least compatible one. On a host with `sm_90` and `sm_120`, torch 2.4.1/CUDA 12.4 passes the `sm_90` floor and produces no finding even though operations on `sm_120` will fail. A mixed-GPU regression test would help cover this.

File: pkg/doctor/check_env_gpu_compat.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+func (c *GPUCompatibilityCheck) Check(ctx *CheckContext) ([]Finding, error) {
+	if os.Getenv("COG_SKIP_GPU_CHECK") != "" {
+		return nil, nil
+	}
+	if ctx.Config == nil || !ctx.Config.Build.GPU {
+		return nil, nil
+	}
+
+	torchVersion, hasTorch := ctx.Config.TorchVersion()
```
Comment (anish-sahoo) [inline pkg/doctor/check_env_gpu_compat.go]: `TorchVersion()` does not establish that this is an exact pin: the current requirement parsing discards the comparator and returns the first version token. As a result, `torch<2.7.0` is treated as `2.7.0`, so this check can pass even though pip installs an older, incompatible release. Could we evaluate only normalized exact `==` pins and skip ranges, direct URLs, and unresolved requirements? This should also have regression coverage.

File: pkg/doctor/check_env_gpu_compat.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+		// Older than any floor we know; say nothing rather than guess.
+		return nil
+	}
+
+	if torchVersion == "" {
+		// torch listed but unpinned; pip resolves a current release. Nothing to compare.
+		return nil
+	}
+
+	torchOK := version.GreaterOrEqual(torchVersion, floor.MinTorch)
```
Comment (anish-sahoo) [inline pkg/doctor/check_env_gpu_compat.go]: `version.GreaterOrEqual` includes local metadata in equality, so `2.7.0+cu128` is not considered greater than or equal to `2.7.0`: the release is not greater, and equality fails because the metadata differs. That produces a false warning at the exact supported floor. Could we strip the local modifier for the torch release comparison, retain it separately for CUDA resolution, and add an exact-boundary test?

File: pkg/doctor/check_env_gpu_compat_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	// Complete populates the requirements content that TorchVersion reads; without it the
+	// check skips before reaching the comparison.
+	require.NoError(t, cfg.Complete(t.TempDir()))
+	return &CheckContext{ctx: context.Background(), ProjectDir: t.TempDir(), Config: cfg}
+}
+
+func TestGPUCompatibilityCheck_RunsWithoutError(t *testing.T) {
+	// A GPU may or may not be present in the test environment; just ensure no panic or error.
+	ctx := gpuContext(t, true, "2.4.1", "12.4")
+	_, err := (&GPUCompatibilityCheck{}).Check(ctx)
```
Comment (anish-sahoo) [inline pkg/doctor/check_env_gpu_compat_test.go]: This test only verifies that the check returns no error and ignores its findings, so the production detection path is not exercised deterministically. Could we test it with a fake `nvidia-smi` and cover multiple GPUs, malformed output, command failure, and registration through the doctor runner? That would also catch the mixed-GPU issue noted above.

File: pkg/doctor/gpu_compatibility.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+// a build can succeed and produce an image where every CUDA operation fails at runtime with
+// "no kernel image is available for execution on the device".
+type GPUCompatibilityCheck struct{}
+
+func (c *GPUCompatibilityCheck) Name() string        { return "env-gpu-compat" }
+func (c *GPUCompatibilityCheck) Group() Group        { return GroupEnvironment }
+func (c *GPUCompatibilityCheck) Description() string { return "GPU compatibility" }
+
+func (c *GPUCompatibilityCheck) Check(ctx *CheckContext) ([]Finding, error) {
+	if os.Getenv("COG_SKIP_GPU_CHECK") != "" {
```
Comment (anish-sahoo) [inline pkg/doctor/gpu_compatibility.go:61]: Could we document `COG_SKIP_GPU_CHECK` in `docs/environment.md` alongside `COG_SKIP_DOCKER_CHECK`, then regenerate `docs/llms.txt`? This is a new user-facing environment variable and its intended use when building for different hardware should be discoverable outside the source.

File: pkg/doctor/gpu_compatibility.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
+package doctor
```
Comment (anish-sahoo) [inline pkg/doctor/gpu_compatibility.go:1]: Could we rename these files to `gpu_compatibility.go` and `gpu_compatibility_test.go`? We prefer subject-based filenames here and spell out `compatibility` rather than shortening it to `compat`.

**Review by anish-sahoo — CHANGES_REQUESTED** [review]
Thanks for putting this together. This is a useful diagnostic and the overall approach looks promising. I found several compatibility edge cases, along with some test and documentation gaps, that need additional fixes before we can work toward merging this. Details are inline.

File: pkg/doctor/gpu_compatibility.go (hunk 6 above)
Comment (golithe) [reply pkg/doctor/gpu_compatibility.go:1]: done

File: pkg/doctor/gpu_compatibility.go (hunk 5 above)
Comment (golithe) [reply pkg/doctor/gpu_compatibility.go:61]: done

File: pkg/doctor/check_env_gpu_compat.go (hunk 1 above)
Comment (golithe) [reply pkg/doctor/check_env_gpu_compat.go]: the check now evaluates every distinct capability rather than only the lowest. your described scenario (sm_90 + sm_120 with torch 2.4.1 is the regression test `TestGPUCompatibilityCheck_MixedGPUsFireForStrictestFloor`

wdyt?

File: pkg/doctor/check_env_gpu_compat.go (hunk 2 above)
Comment (golithe) [reply pkg/doctor/check_env_gpu_compat.go]: - added `requirements.ExactVersion` and `Config.TorchExactVersion` where only a single `==` specifier counts
- ranges/wildcards/direct URLs/unpinned all produce no finding rather than a guess
- tested in `TestExactVersion` and `TestGPUCompatibilityCheck_SkipsNonExactPin`

File: pkg/doctor/check_env_gpu_compat.go (hunk 3 above)
Comment (golithe) [reply pkg/doctor/check_env_gpu_compat.go]: - the torch comparison now strips the local tag before `version.GreaterOrEqual`, so 2.7.0+cu128 passes at the exact floor
- CUDA bound is still checked separtely from Build.CUDA
- boundari case in `TestEvaluateGPUCompat` ("sm_120 exact floor with local tag passes)

File: pkg/doctor/check_env_gpu_compat_test.go (hunk 4 above)
Comment (golithe) [reply pkg/doctor/check_env_gpu_compat_test.go]: - replaced the `RunsWithoutError` test entirely
- the nvidia-smi probe is now a package-level functin variable that tests stub
- went with that over fake bin on PATH to keep it hermetic, same intent
- now covers multiple GPUs, command failure, malformed output, non-exact pins, and runner registration

**Comment by golithe** [comment]
@anish-sahoo ty for your thorough review! i addressed your comments, PTAL

**Review by anish-sahoo — CHANGES_REQUESTED** [review]
Thanks for addressing the first review round. The mixed-GPU evaluation, exact-pin handling for ordinary `torch==...` requirements, local-tag boundary case, deterministic probe tests, docs, and rename all look good to me.

I found two remaining issues that need to be addressed before merge:

1. `gpu_compatibility.go` strips the local wheel tag and then relies only on `build.cuda`. This can silence a real Blackwell incompatibility: `torch==2.7.0+cu118` with an explicit `cuda: "12.8"` passes the check because it compares the stripped `2.7.0` and base CUDA 12.8 against the floor. But Cog retains the `+cu118` wheel, and that wheel has no Blackwell kernels regardless of its CUDA base image. Please preserve/derive the wheel CUDA tag for compatibility evaluation, rather than losing it when normalizing the torch release, and add a full `Config.Complete -> Check` regression test for this override combination.

2. `Config.TorchExactVersion()` compares `requirements.PackageName(pkg)` directly to lowercase `"torch"`. Exact valid requirements such as `torch[extra]==2.4.1` and `Torch==2.4.1` therefore return no torch pin and skip the GPU check. Please normalize the base distribution name (strip extras and normalize casing, ideally PEP 503-style) before matching, with an end-to-end check regression test.

**Comment by golithe** [comment]
> `gpu_compatibility.go` strips the local wheel tag and then relies only on `build.cuda`.

good catch. now derives the wheel's CUDA from the `+cuXXX` local tag (via the exported `config.CUDAVersionFromTorchLocalTag`) and compares that against the floor instead of `build.cuda`, so `torch==2.7.0+cu118` with `cuda: "12.8"` fires. regression: `TestGPUCompatibilityCheck_LocalCUDATagOverridesBuildCUDA`

> `Config.TorchExactVersion()` compares `requirements.PackageName(pkg)` directly to lowercase `"torch"`.

fixed. added `requirements.NormalizePackageName` (PEP 503: strip extras, collapse -_., lowercase) and match on it in` pythonPackageExactVersion`, so `torch[extra]==2.4.1` and `Torch==2.4.1` resolve instead of being skipped. Regression: `TestNormalizePackageName` + end-to-end `TestGPUCompatibilityCheck_NormalizesTorchName`.

**Comment by golithe** [comment]
@anish-sahoo adressed both, PTAL

File: pkg/doctor/gpu_compatibility.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+	if torchVersion == "" {
+		// torch listed but unpinned; pip resolves a current release. Nothing to compare.
+		return nil
+	}
+
+	// A wheel's +cu118 tag is authoritative over the build's base CUDA: torch==2.7.0+cu118
+	// ships no Blackwell kernels even under cuda: "12.8", so evaluate against 11.8, not 12.8.
+	// Fall back to the build CUDA only when there is no cuXYZ tag to derive from.
+	effectiveCUDA := cuda
+	if wheelCUDA, _ := config.CUDAVersionFromTorchLocalTag(torchVersion); wheelCUDA != "" {
```
Comment (anish-sahoo) [inline pkg/doctor/gpu_compatibility.go]: I re-tested this and need to correct my earlier assumption. The GPU check and the build resolve CUDA differently:

- `torch==2.7.0+cu118` with `cuda: "12.8"` builds with the cu128 wheel, but the check uses cu118 and warns incorrectly.
- An explicit cu118 index builds with the cu118 wheel, but the check uses `cuda: "12.8"` and does not warn.

Could the check use the final resolved requirement and index from `PythonRequirementsForArch`? That would ensure it checks the wheel Cog will actually install.

**Review by anish-sahoo — CHANGES_REQUESTED** [review]
The GPU compatibility check can evaluate a different wheel than Cog actually installs, which causes both false positives and false negatives. Everything else LGTM.

File: pkg/doctor/gpu_compatibility.go (hunk 7 above)
Comment (golithe) [reply pkg/doctor/gpu_compatibility.go]: good catch again, and you're right. 

- the check now resolves the wheel via the same path as `PythonRequirementsForArch` (new `Config.ResolvedTorchWheel`) and derives CUDA from the resolved package index instead of the requirement's local tag, so it matches what Cog installs in both directions: `torch==2.7.0+cu118` + `cuda: "12.8"` resolves to cu128 and stays silent, and an explicit `--extra-index-url=.../cu118` fires on 11.8. 
- Regressions for both in `TestResolvedTorchWheel` and the two `GPUCompatibilityCheck` tests. PTAL.

**Review by anish-sahoo — APPROVED** [review]
Thank you for contributing. LGTM!

**Comment by golithe** [comment]
> Thank you for contributing. LGTM!

Thank you for reviewing! cheers

**Comment by anish-sahoo** [comment]
note to @golithe - this will be merged soon, we are a bit swamped right now, I'm waiting for one of the repo admins to override the codeql check.

**Comment by golithe** [comment]
hi @anish-sahoo, is there any action required from my side regarding the failing CI? i rebased on main to have a clear diff, the failing CI steps mention too many requests.

cheers

**Review by michaeldwan — APPROVED** [review]
nice, thanks!

**Comment by michaeldwan** [comment]
@golithe GH is stuck waiting on CodeQL checks that passed weeks ago. Going to push an empty commit to trigger it and get this merged. Sorry for the wait!

**Comment by golithe** [comment]
@michaeldwan thanks!
